addons/t9n/models/message.py

```diff
@@ -11,14 +11,67 @@ class Message(models.Model):
     _description = "Localizable message"
 
     body = fields.Text(
+        "Entry to be Translated",
         help="The actual, textual content to be translated.",
     )
+    context = fields.Char(
+        help="Context of the message.",
+    )
+    translator_comments = fields.Text(
+        help="Comments wrote by the translator/developer in the resource file.",
+    )
+    extracted_comments = fields.Text(
+        help="Comments extracted during the process of extracting translatable strings.",
+    )
+    references = fields.Text(
+        help="The full text that represents the references, one per line.",
+    )
     resource_id = fields.Many2one(
         comodel_name="t9n.resource",
         help="The resource (typically a file) from which the entry is coming from.",
+        ondelete="cascade",
+        required=True,
     )
     translation_ids = fields.One2many(
         comodel_name="t9n.translation",
         inverse_name="source_id",
         string="Translations",
     )
+
+    _sql_constraints = [
+        (
+            "body_context_resource_unique",
+            "UNIQUE(body, context, resource_id)",
+            "The combination of the message body and context must be unique within the same resource!"
+        ),
+    ]
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        updated_records = self.env["t9n.message"]
+
+        resource_ids = list(set(vals.get("resource_id") for vals in vals_list if "resource_id" in vals))
+        resources = self.env["t9n.resource"].browse(resource_ids)
+        project_resources_map = {res.id: res.project_id for res in resources}
+
+        bodies = set(vals.get("body") for vals in vals_list)
+        contexts = set(vals.get("context") for vals in vals_list)
+
+        potential_matches = self.search([
+            ("body", "in", list(bodies)),
+            ("context", "in", list(contexts))
+        ])
+
+        for vals in vals_list:
+            cur_project_id = project_resources_map.get(vals.get("resource_id"))
+
+            matching_records = potential_matches.filtered(lambda r: r["body"] == vals.get(
+                "body") and ((r["context"] == vals.get("context")) or (not r["context"] and vals.get("context") is None)) and r["resource_id"].project_id == cur_project_id)
+
+            if matching_records:
+                existing_record = matching_records[0]
+                existing_record.write(vals)
+                updated_records += existing_record
+            else:
+                updated_records += super().create([vals])
```

Comment:
does += work for records? The standard is |=, but be careful because it can be a bottleneck for large number of records. Sometimes we store the ids in a set and then do a browse of all the ids to avoid it.

Also, a matter of preference, but you can use continue to avoid using an `else` statement (and have it be the default behavior if we add in additional conditions later on)

```suggestion
            if matching_records:
                existing_record = matching_records[0]
                existing_record.write(vals)
                updated_records |= existing_record
                continue
            updated_records |= super().create([vals])
```

Reply:
plus `+=` doesn't prevent duplicates 👆

Reply:
neither does `|=` :sweat_smile:  that's partially why we use a set and collect the IDs sometimes

Reply:
I had to look it up, because it would be super counterintuitive that `|=` would allow duplicates, since it's a logical OR operation. From what I found, the union function in a RecordSet does use an `OrderedSet` internally for the ids when using the operator, so it does prevent duplicates.

![image](https://github.com/odoo-dev/odoo/assets/5985577/965c2693-3e37-47b0-8052-7f0cf0a1ae36)

Reply:
![image](https://github.com/odoo-dev/odoo/assets/20225337/9753dfe0-660d-40dd-8701-ec9633132a43)